hare/algorithm.go

```diff
@@ -161,25 +163,60 @@ func roleFromRoundCounter(k uint32) Role {
 	}
 }
 
-func (proc *ConsensusProcess) handleMessage(m *pb.HareMessage) {
-	// Note: instanceId is already verified by the broker
+func (proc *ConsensusProcess) onEarlyMessage(m *pb.HareMessage) {
+	pub, err := crypto.NewPublicKey(m.PubKey)
+	if err != nil {
+		log.Warning("Could not construct public key: ", err.Error())
+		return
+	}
 
-	// validate message
-	if !proc.validator.ValidateMessage(m, proc.k) {
-		log.Warning("Message is not syntactically valid")
+	if _, exist := proc.pending[pub.String()]; exist { // ignore, already received
+		log.Warning("Already received message from sender %v", pub.String())
 		return
 	}
 
-	// TODO: validate claimedRole proof
+	proc.pending[pub.String()] = m
+}
+
+func (proc *ConsensusProcess) validateRole(m *pb.HareMessage) bool {
+	if m.Message == nil {
+		log.Warning("Role validation failed: message is nil")
+		return false
+	}
+
+	// TODO: validate role proof
 
 	// validate claimedRole
 	expectedRole := proc.oracle.Role(m.Message.K, Signature(m.Message.RoleProof))
 	claimedRole := roleFromRoundCounter(m.Message.K)
 	if expectedRole != claimedRole {
 		log.Warning("Invalid claimedRole detected. Expected: %v Actual: %v", expectedRole, claimedRole)
+		return false
+	}
+
+	return true
+}
+
+func (proc *ConsensusProcess) handleMessage(m *pb.HareMessage) {
+	// Note: instanceId is already verified by the broker
+
+	// first validate role
+	if !proc.validateRole(m) {
+		log.Warning("Role validation failed")
 		return
 	}
 
+	// validate message for this or next round
+	if !proc.validator.ValidateMessage(m, proc.k) {
+		if !proc.validator.ValidateMessage(m, proc.k+1) {
+			log.Warning("Message is not syntactically valid for either round")
```

Comment:
The message isn't valid, not necessarily syntactically valid

Reply:
Fixed